doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+  `restoreadminfromsnapshot` steps, respectively.
+
+* Create an additional OpenStack controller or compute node as a scratch node:
+  this will allow you to deploy only your own barclamp's controller side
+  components to this node while all other OpenStack services reside on the main
+  controller. This greatly shortens the time a chef run for your barclamp's
+  controller side components takes and thus speeds up development. Make a
+  backup of this node as well to guard against breakage and allow for
+  clean-slate deployment of your barclamp's controller side components.
+
+The best way to set up your scratch node is to make it a lonely node. To this
+end, you will need to re-run mkcloud after you have created your cloud:
+
+```
+nodenumberlonelynode=1 mkcloud setuplonelynodes crowbar_register
+
+```
+
+## Development Environment
+
+Most of your development will take place on your Crowbar node (log in as root).
+You will find the currently running Crowbar code (both Rails application
+components and chef recipes) in the `/opt/dell` directory. We will provide
+detailed locations for all important components of a Barclamp in the _Basic
+Implementation_ section below.
+
+In order to track changes you make we recommend you keep a checkout of your
+`crowbar-openstack` fork on the Crowbar node and apply the changes you make in
+there
+
+### Preparations for Crowbar Node
+
+These are recommendations, but we'll assume you followed them for the remainder
+of this document since this shortens the commands in our examples. First of
+all, clone your `crowbar-openstack` fork to `/root/crowbar-openstack` on your
+Crowbar node and switch to your development branch.
+
+Next, create the following /root/.bashrc with a few useful aliases and
+environment variables (substitute with appropriate values where so indicated by
+comments):
+
+```
+# Replace "Crowbar Committer" by your name as it should appear in commit
+# messages.
+C_NAME="Crowbar Committer"
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by the ID of the oldest
+# commit on your branch. Running
+#
+#   git log --format='%H %s' $branch..HEAD
+#
+# where $branch is the upstream branch you are developing against such as
+# `master` or `stable/3.0`, should give you a list of original commits. This
+# assumes $branch has not been improperly cherry-picked into or rebased in your
+# fork, though. It also assumes that you are not working on your fork's
+# `master` branch (you shouldn't, but people have been known to do this
+# anyway). Since all of these may occasionally happen, this guide leaves it up
+# to the user to set $COMMIT as they see fit rather than trying to determine
+# its value automatically through the command mentioned above.
+
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
```

Comment:
There may be multiple commits in a topic branch. Not everyone uses the OpenStack model of `git commit --amend` that leaves just a single commit in a topic branch. It doesn't make sense without Gerrit or similar heavy CI artillery. Likewise, `master` or `stable/3.0` may be way out of date in a fork (it commonly is). The raw SHA1 guarantees you a diff from your first commit to HEAD on the topic branch. Nothing more, nothing less. That is why I opted for this.